### DISCUSSION TITLE
Alternative way of coding

### DISCUSSION THREAD

**Discussion by hohse** [openqda/openqda#143] [Ideas]
When presenting OpenQDA at the DGPuK2024 and even before when discussing it with students, we got several hints to create one more way to code text: The idea would be to be able to "activate" a code and then every time a snipped of text gets selected/highlighted, this activated code gets applied to it. So for example, you activate a code called "place", and then go through a whole text document, highlight "places" and they get instantly coded accordingly.

**Comment by jankapunkt** [comment]
You mean, like the "text marker" from several text processors?

**Reply by hohse** [reply]
Yeah, that's one way to imagine it: The activated code would work like a text marker then.

**Reply by kadewe** [reply]
I like this idea, it speeds up the work. Especially in a first walk through a text, people use a code "Interesting quotes" as a first orientation. So I imagine ACTIVATING (UI) the CODE e.g. "Interesting quotes" and then just highlight all the interesting parts to code. 
Needs to be discussed with UX, how to switch the code into ACTIVATED and not-ACTIVATED and how to show this to the users (e.g. activated code highlighted, others are dimmed and not clickable until you de-activate the activated code - or if you click on the non-activated codes it asks you you de-activate the activated code)

**Comment by matzefrey** [comment]
Another very practical and fast way of coding would be to assign shortcuts to codes. But I'm not sure if that is possible in a browser environment. 

However, thanks a lot for creating this project! I'm really looking forward to use productively!

**Reply by kadewe** [reply]
Well, keyboard shortcuts are easily available in a browser https://hackernoon.com/add-keyboard-shortcuts-to-your-web-app-ba358016ff05 - but we need to think this through from an UX point of view:
(1) you select a text within the textview-pane.
(2a) you press some shortcut key and the selected text is coded with the corresponding code 
- OR - 
(2b) you press the coding-shortcut key, the available shortcuts are displayed and you choose them with your up-down-keys or can search-filter them?

Let's start with (2a) - it is easier - you can assign up to ten shortcut keys (0-9) to your codes. Then you just can select some text, hit "2" and the text is coded with the assigned code.
UI-Problem: how do you assign the key? In the Code-menu. how do you show the shortcut key? Besides the code button.

**Reply by kadewe** [reply]
let's have a look at this again. How about automated assignment of the first 10 codes you use to a shortcut key from 0-9. if they are full, you can assign a number to another code. the numbers are shown in the code tree. 

and even have a few more shortcuts in coding view: 
v - create in-vivo-code from selection
n - create new code
r - code the current selection with the last code
what else comes to your mind if you think about keyboard shortcuts?

**Reply by matzefrey** [reply]
Thanks for picking this up! 
That sounds reasonable. And I like the three additional shortcuts. However, “n” is on the right side of the keyboard. So if I'm using my right hand to operate the mouse and want to create a new code (n), I either have to switch hands or move my right hand from the mouse to the keyboard. 

Maybe this is completely irrelevant in practice, but right now I think it would make more sense to move this shortcut to the left side as well. 

In addition: would it be theoretically possible to customize the shortcuts myself?

**Reply by anouarg88** [reply]
Hi! Just wanted to say that this sounds like a cool idea and chip in a thought: I would love the option to trash the mouse altogether and mark the text without your hands having to leave the keyboard. I just tried and found that you can mark sections using shift+cursor as you would in any text editor. However, there is no option to move through speaker turns/lines with the keyboard. Also, the cursor is not easily visible until you mark something (if you're actually on the right page). I might be part of a backward minority in thinking that using the mouse is tedious, but maybe this would be something worth considering?

By the way, I think 2a and 2b both sound like something I would need. Maybe instead of shortcuts statically assigned to codes, they could be assigned in a scrollable area (PgUp/PgDwn), so 0-9 would apply to whichever codes you've been working on most recently? I think this fits the current logic, since codes are already assigned to the bottom of the code list. But if you have a huge codebook, then scrolling through all those pages to find your code is a nuisance, so a search function would be useful.
Also: 
u - unassign
m - memo

**Reply by kadewe** [reply]
Excellent observation with "n" being a bit to the right – i am typing 10-finger-system, so I may say that "b" is still left index finger and "n" is like an octave stretch – possible but still a stretch. So maybe "c" (create) or "a" (add) would be better?
I also like u and m! Also I think we should look at "old school" keyboard based editors such as emacs and vim, because they probably have a few tricks up their sleeve – anyone wiling to share his or her experiences for how to do keyboard based text navigation and selection?
